### PR TITLE
Remove cancancan customizations

### PR DESCRIPTION
**Pull request by coorasse**
**Description**
While testing solidus against cancancan 3.0 I noticed that the usage was "not correct (*Euphemismus*)".
This PR aims to remove the customizations introduced to cancancan, which are not necessary.
The current customizations are not compatible with version 3.0.0 of cancancan since Solidus makes use of `merge` of Abilities. The `3.0.0` has a breaking change which changes how `aliased_actions` are merged.

I removed all custom aliases and replaced `display` with `read`. Also, I corrected the rule which would have caused issues during the migration.

This is probably a breaking change, because it will cause problems for the plugins.
Plugins should now use standard `cancancan` actions and replace `display` with `read`.

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [x] I have updated Guides and README accordingly to this change (if needed)
- [x] I have added tests to cover this change (if needed)

### CODE REVIEW

File: api/app/controllers/spree/api/base_controller.rb [hunk 1]
```diff
@@ -133,13 +133,13 @@ def find_product(id)
 
       def product_scope
         if can?(:admin, Spree::Product)
-          scope = Spree::Product.with_deleted.accessible_by(current_ability, :read).includes(*product_includes)
+          scope = Spree::Product.with_deleted.accessible_by(current_ability).includes(*product_includes)
```
Comment (coorasse) [inline api/app/controllers/spree/api/base_controller.rb]: this is not necessary anymore. The default is `:index` which is by default aliased by `:read`.

File: api/app/controllers/spree/api/orders_controller.rb [hunk 2]
```diff
@@ -49,8 +49,9 @@ def empty
         render plain: nil, status: 204
       end
 
+      # TODO: this action is never accessible. shall we simply remove it?
```
Comment (coorasse) [inline api/app/controllers/spree/api/orders_controller.rb]: this method was never accessible. why don't we remove it?

File: core/app/models/spree/ability.rb [hunk 3]
```diff
@@ -29,25 +29,12 @@ def self.remove_ability(ability)
...
       activate_permission_sets
       register_extension_abilities
     end
 
     private
 
-    def alias_actions
-      clear_aliased_actions
-
-      # override cancan default aliasing (we don't want to differentiate between read and index)
```
Comment (coorasse) [inline core/app/models/spree/ability.rb]: this is supported already. you can use `read`.

File: core/app/models/spree/ability.rb [hunk 4]
```diff
@@ -29,25 +29,12 @@ def self.remove_ability(ability)
...
-      alias_action :delete, to: :destroy
```
Comment (coorasse) [inline core/app/models/spree/ability.rb]: `delete` was never used

File: core/app/models/spree/ability.rb [hunk 5]
```diff
@@ -29,25 +29,12 @@ def self.remove_ability(ability)
...
-      alias_action :edit, to: :update
-      alias_action :new, to: :create
-      alias_action :new_action, to: :create
```
Comment (coorasse) [inline core/app/models/spree/ability.rb]: `new_action` was never used

File: core/lib/spree/permission_sets/default_customer.rb [hunk 6]
```diff
@@ -4,11 +4,11 @@ module Spree
...
       def activate!
-        can :display, Country
-        can :display, OptionType
-        can :display, OptionValue
+        can :read, Country
+        can :read, OptionType
+        can :read, OptionValue
         can :create, Order
-        can [:read, :update], Order do |order, token|
+        can [:show, :edit, :update], Order do |order, token|
```
Comment (coorasse) [inline core/lib/spree/permission_sets/default_customer.rb]: block conditions should be defined only for actions on object instances, not classes, therefore this rule applies only on `:show, :edit, :update`. This was actually working by mistake since version 2.0 of cacancan

File: core/spec/models/spree/ability_spec.rb [hunk 7]
```diff
@@ -254,7 +256,7 @@ def initialize(_user)
       context 'requested by same user' do
         let(:resource) { user }
         it_should_behave_like 'access granted'
-        it_should_behave_like 'no index allowed'
+        it_should_behave_like 'index allowed'
```
Comment (coorasse) [inline core/spec/models/spree/ability_spec.rb]: why should it not be allowed?

**Comment by kennyadsl** [comment]
@coorasse wow, thank you, I love this change! Do you think there's a way to deprecate the old usage any way before removing it?

**Comment by coorasse** [comment]
I cannot really think of a deprecation warning: we could override `can?` and `accessible_by` methods of `cancancan` and raise a warning when `can?` is called with `delete`, `new_action`, `read` or `display` and when `accessible_by` is called with `:read` or `:display`.

The calls to `accessible_by(ability, :read)` should be replaced by `accessible_by(ability)` but this would not work before this update! ⚠️ 

Also `can? :read, resource` should be changed to the specific action `can? :show` or `can? :edit`, etc, but again: it would not work before this change.

The calls to `can? :delete, resource` could already be replaced by `can? :destroy, resource`.

Also the calls to `can? :new_action, resource` could already be replaced by `can? :create, resource`.

The main problem was this alias:
```
alias_action :show, to: :read
alias_action :index, :read, to: :display
```
which caused `read` to be the same of `show` and `display` to be both `read` and `index`...
I am not even sure what this means 😄. If you notice, in fact, the changes I had to do in the code itself were not actually many. It is more a renaming issue. I can make a new comment and show in detail how/why the problem is raised only in version `3.0.0`.

I let you guys decide about how to move here: I have not enough knowledge about the project.
I pointed the finger 😄 and I can provide all help you need but I think this might break quite some stuff.

**Comment by coorasse** [comment]
In cancancan 3.0.0 when two abilities are merged, also the aliased_actions are merged.
Solidus defines the following aliases in base ability:
```
{
:destroy=>[:delete], 
:update=>[:edit], 
:create=>[:new, :new_action], 
:read=>[:show], 
:display=>[:index, :read]
}
```

base ability is the merged with other abilities where the default aliases are defined. The following:

```
{
  :read => [:index, :show],
  :create => [:new],
  :update => [:edit]
}
```

with the following aliases as a result:

```
{
:destroy=>[:delete], 
:update=>[:edit], 
:create=>[:new], 
:read=>[:index, :show], 
:display=>[:index, :read]
}
```

The call that Solidus performs on `Spree::Order.accessible_by(ability, :read)` in version 3.0.0 fails because the following rule:
```
can [:read, :update], Order do |order, token|
  order.user == user || (order.guest_token.present? && token == order.guest_token)
end
```
 is now identified as possible interesting rule and cancancan raises an issue because this is a block rule (`do...end`) defined on a collection action `read`

File: api/app/controllers/spree/api/base_controller.rb (hunk 1 above)
Comment (jkojro) [reply api/app/controllers/spree/api/base_controller.rb]: Hi @coorasse! But if `:read` alias was `:show` shouldn't it be changed like this `scope = Spree::Product.with_deleted.accessible_by(current_ability, :show).includes(*product_includes)` to keep same logic? Isn't your change changing scope from show to index now?

Comment (coorasse) [reply api/app/controllers/spree/api/base_controller.rb]: it changes from `:show` to `:index` but all the rules are defined for `:read` instead of `:display` which was including also `:read`. it's quite complicated, I know, I don't know why all these aliases were generated before.  `accessible_by(ability, :show)` would not make sense because you want a list of objects. `:show` is for single instances.

Comment (jkojro) [reply api/app/controllers/spree/api/base_controller.rb]: Yeah, that make sense. So for now, when is `alias_action :show, to: :read` this scope with argument `(current_ability, :read)` had no authorization at all I gues?

File: api/app/controllers/spree/api/orders_controller.rb (hunk 2 above)
Comment (jkojro) [reply api/app/controllers/spree/api/orders_controller.rb]: @coorasse Could you explain why is it unaccessible?

Comment (coorasse) [reply api/app/controllers/spree/api/orders_controller.rb]: it was defined with `:index` and since `:read` was an alias of `:show`, no rule was defined for `:index`. in fact, I replaced it with a random action called `:impossible_action` and no test failed. There is only one test and it checks that this endpoint should not be accessible 🤷‍♂️ . That's why I asked myself: "why don't we remove it?"

File: api/app/controllers/spree/api/base_controller.rb (hunk 1 above)
Comment (coorasse) [reply api/app/controllers/spree/api/base_controller.rb]: `:read` was aliased by `:display` and rules were (almost) always defined on `:display`. You can see that most of the changes are about that: https://github.com/solidusio/solidus/pull/3148/files#diff-82b30f937458e7c1052c326354dc18cbL8

File: api/app/controllers/spree/api/orders_controller.rb (hunk 2 above)
Comment (jkojro) [reply api/app/controllers/spree/api/orders_controller.rb]: @coorasse As I've tested, `:manage` gives acces to all rules, also theese not defined, so a user with `can :manage, Order` will go through `authorize! :impossible_action, Order` too. So it could be replaced with `authorize! :admin, Order` I think and it would keep eligible logic.

File: api/app/controllers/spree/api/orders_controller.rb [hunk 8]
```diff
@@ -49,8 +49,9 @@ def empty
...
       def index
-        authorize! :index, Order
```
Comment (jkojro) [inline api/app/controllers/spree/api/orders_controller.rb]: `authorize! :admin, Order`

File: api/app/controllers/spree/api/orders_controller.rb (hunk 2 above)
Comment (coorasse) [reply api/app/controllers/spree/api/orders_controller.rb]: Oh right! `manage` authorizes you! I forgot that 🤦‍♂️ . To do it properly here we should leave `authorize! :index, Order` but then [this test](https://github.com/solidusio/solidus/blob/29fe2cc432ac336ee4dd407f677512c28b35c001/api/spec/requests/spree/api/orders_controller_spec.rb#L201) will fail. So I'd say we could use `authorize! :admin, Order` and keep the current behaviour.

File: api/app/controllers/spree/api/base_controller.rb (hunk 1 above)
Comment (tvdeyen) [reply api/app/controllers/spree/api/base_controller.rb]: I agree that `:index` is the right thing here, as `:show` would need to check every single instance of a Product if its able to be shown to the user, what is definitely not what we want here.

File: api/app/controllers/spree/api/orders_controller.rb [hunk 9]
```diff
@@ -50,7 +50,7 @@ def empty
       end
 
       def index
-        authorize! :index, Order
+        authorize! :admin, Order
```
Comment (tvdeyen) [inline api/app/controllers/spree/api/orders_controller.rb]: Why this change?

**Comment by kennyadsl** [comment]
Closing in favor of #3701.
